### DISCUSSION TITLE
wallet_switchEthereumChain not working

### DISCUSSION THREAD

**Discussion by Pasha8914** [WalletConnect/walletconnect-monorepo#2456] [Developer Support]
stack:
> react 18.2.0
> @walletconnect/ethereum-provider ^2.7.5
> @web3modal/standalone ^2.4.1

When I change the chain, it changes in the provider, in the signer, but not in the wallet (trust wallet)
```
class WalletConnector {
  public provider: Provider | undefined

  public async setupProvider() {
      try {
        this.provider = await Provider.init( {
            projectId,
            showQrModal: true,
            chains: [1,137],
            metadata: {
              name: protocol.name,
              url: protocol.appUrl,
              icons: [protocol.logoUrl],
              description: protocol.description,
            },
        })
        return this.provider
      } catch (err) {
        throw new Error(`WalletConnect: ${err.message}`)
      }
  }
}

const connector = new WalletConnector()

export function useWalletProvider() {
    const [provider, setProvider] = useState<Provider | undefined>(undefined)
    
    const getProvider = useCallback(() => {
        if (!connector.provider?.session) {
          await connector.setupProvider()
        }
        return connector.provider
    }, [connector])
    
    useEffect(() => {
      getProvider(chainId).then(_provider =>setProvider(_provider) )
    }, [])
    
    return provider
}

export function useChangeNetwork() {
  const provider = useWalletProvider()

  return useCallback(
    async (chainId: SupportedChainId) => {
      if (!provider) {
        throw new Error(errors.wallet.WALLET_REQUIRED)
      }
      try {
        const params = [{ chainId: ADD_NETWORK_LIST[chainId].chainId }]

        await provider.request({ method: 'wallet_switchEthereumChain', params })
      } catch (e) {
        if (e.code === 4902 || e.message.includes('wallet_addEthereumChain')) {
          await provider.sendRequest({
            method: 'wallet_addEthereumChain',
            params: [ADD_NETWORK_LIST[chainId]],
          })
        }
      }
    },
    [provider],
  )
}

**Comment by glitch-txs** [comment]
if you want to change to polygon try:
`provider.signer.setDefaultChain('eip155:137')`
in WalletConnect v2 you actually connect to multiple chains at the same time, setting a chain to default is handy to choose the network for the transaction, but you don't actually expect the wallet to request a change of chain to the user since it's already connected to all the first selected ones at the same time.

**Reply by Pasha8914** [reply]
> if you want to change to polygon try: `provider.signer.setDefaultChain('eip155:137')` in WalletConnect v2 you actually connect to multiple chains at the same time, setting a chain to default is handy to choose the network for the transaction, but you don't actually expect the wallet to request a change of chain to the user since it's already connected to all the first selected ones at the same time.

all fields changed in the provider object, but the trust wallet still submits transactions for signature in the wrong network

**Reply by glitch-txs** [reply]
How are you triggering the tx? You need to change the default chain in the EthereumProvider first before wrapping it with any library like ethers.js, since this libraries currently don't support multi-chain

**Reply by glitch-txs** [reply]
Adding to this, the way you switch chain should work a well, just saw it is abstracted in the library 👍

**Reply by Pasha8914** [reply]
> Adding to this, the way you switch chain should work a well, just saw it is abstracted in the library 👍

```
// AbstractWalletProvider

 private hexlifyParams(params: Params | Record<string, string>) {
    const result = Object.assign({}, params)

    const numericParams: Array<keyof Params> = [
      'gas',
      'type',
      'nonce',
      'value',
      'gasPrice',
      'maxFeePerGas',
      'maxPriorityFeePerGas',
    ]

    numericParams.forEach((key) => {
      const value = params[key]
      if (value) {
        result[key] = hexValue(value)
      }
    })

    return result
  }
  
  
private prepareRequest({ method, params }: SendRequestParams) {
    switch (method) {
      case 'eth_call':
      case 'eth_sendTransaction': {
        if (params instanceof Array) {
          const [args] = params
          if (args instanceof Object) {
            //TODO fix SendRequestParams

            return { method, params: [this.hexlifyParams(args as Params)] }
          }
        }
        break
      }
    }

    return { method, params }
  }
  
public async sendRequest<T>(params: SendRequestParams): Promise<T> {
    try {
      const args = this.prepareRequest(params)
      return await this.provider.request<T>(args)
    } catch (err) {
      if (err?.code) {
        throw {
          code: err.code,
          message: `Provider method sendRequest has error: ${err.message}`,
        }
      } else {
        throw new Error(`Provider method sendRequest has error: ${err.message}`)
      }
    }
  }
  
  
  // SEND TX UI METHOD
const sendTx = useCallback(async (_walletProvider: AbstractWalletProvider, _params: Params, gasParams: GasParams) => {
    try {
      return await _walletProvider.sendRequest<string>({
        method: 'eth_sendTransaction',
        params: [{ ..._params, ...gasParams }],
      })
    } catch (err) {
      // https://github.com/MetaMask/metamask-extension/issues/13341
      if (err.message.includes('EIP-1559') && 'maxFeePerGas' in gasParams) {
        return _walletProvider.sendRequest<string>({
          method: 'eth_sendTransaction',
          params: [{ ..._params, gasPrice: gasParams.maxFeePerGas }],
        })
      }
      throw err
    }
  }, [])
  ```

**Reply by Pasha8914** [reply]
> > Adding to this, the way you switch chain should work a well, just saw it is abstracted in the library 👍
> 
> ```
> // AbstractWalletProvider
> 
>  private hexlifyParams(params: Params | Record<string, string>) {
>     const result = Object.assign({}, params)
> 
>     const numericParams: Array<keyof Params> = [
>       'gas',
>       'type',
>       'nonce',
>       'value',
>       'gasPrice',
>       'maxFeePerGas',
>       'maxPriorityFeePerGas',
>     ]
> 
>     numericParams.forEach((key) => {
>       const value = params[key]
>       if (value) {
>         result[key] = hexValue(value)
>       }
>     })
> 
>     return result
>   }
>   
>   
> private prepareRequest({ method, params }: SendRequestParams) {
>     switch (method) {
>       case 'eth_call':
>       case 'eth_sendTransaction': {
>         if (params instanceof Array) {
>           const [args] = params
>           if (args instanceof Object) {
>             //TODO fix SendRequestParams
> 
>             return { method, params: [this.hexlifyParams(args as Params)] }
>           }
>         }
>         break
>       }
>     }
> 
>     return { method, params }
>   }
>   
> public async sendRequest<T>(params: SendRequestParams): Promise<T> {
>     try {
>       const args = this.prepareRequest(params)
>       return await this.provider.request<T>(args)
>     } catch (err) {
>       if (err?.code) {
>         throw {
>           code: err.code,
>           message: `Provider method sendRequest has error: ${err.message}`,
>         }
>       } else {
>         throw new Error(`Provider method sendRequest has error: ${err.message}`)
>       }
>     }
>   }
>   
>   
>   // SEND TX UI METHOD
> const sendTx = useCallback(async (_walletProvider: AbstractWalletProvider, _params: Params, gasParams: GasParams) => {
>     try {
>       return await _walletProvider.sendRequest<string>({
>         method: 'eth_sendTransaction',
>         params: [{ ..._params, ...gasParams }],
>       })
>     } catch (err) {
>       // https://github.com/MetaMask/metamask-extension/issues/13341
>       if (err.message.includes('EIP-1559') && 'maxFeePerGas' in gasParams) {
>         return _walletProvider.sendRequest<string>({
>           method: 'eth_sendTransaction',
>           params: [{ ..._params, gasPrice: gasParams.maxFeePerGas }],
>         })
>       }
>       throw err
>     }
>   }, [])
> ```


I don't use ethers.js or anything, I just get the provider from the connector (in this case it's the walletConnect ethereumProvider) and pass it to the abstract class

**Reply by glitch-txs** [reply]
Mmm I see, I guess you're using react right? As far as I can see I would try to:

1. I see in the getProvider function you're checking for a session, so if the user is not connected you're re-initializing the provider,  I would try to init it just once.

1.  The useWalletProvider hook has a typo on the return value 

About the rest I'm not sure, I've tested it using ethers and worked fine. What do you think it could be @0xAsimetriq ? 👀

**Reply by Pasha8914** [reply]
> Mmm I see, I guess you're using react right? As far as I can see I would try to:
> 
> 1. I see in the getProvider function you're checking for a session, so if the user is not connected you're re-initializing the provider,  I would try to init it just once.
> 2. The useWalletProvider hook has a typo on the return value
> 
> About the rest I'm not sure, I've tested it using ethers and worked fine. What do you think it could be @0xAsimetriq ? 👀

A typo was made when porting the code for the discussion. There are no typos in the project itself. So it cannot affect the operation of the application.

the network has changed in the rps provider, as it should
<img width="562" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/43392281/c8d3ac5a-0028-4df2-b177-8ba7c0563cdf">


but trsut wallet still asks to sign the transaction in the polygon

**Reply by Pasha8914** [reply]
also i got error in console

No matching key. keychain: 1babde695c130e5c76c6881dada7b542c8fa6cfb6a5d88f1d984ad865e95b969

<img width="582" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/43392281/073cf813-82b9-46ec-aa1c-2430f6649dcf">

**Reply by glitch-txs** [reply]
Do you think you can create a minimal reproduction repository of it? I can try help you debug

**Reply by Pasha8914** [reply]
> Do you think you can create a minimal reproduction repository of it? I can try help you debug

sure, i will do

**Reply by Pasha8914** [reply]
@glitch-txs 
I invited you to the test repository

**Reply by glitch-txs** [reply]
hi @Pasha8914 I just ran your example and works perfectly for me, I'm able to send txs both for Polygon and Mainnet, which wallet are you using? I tested it with Trust Wallet.

**Reply by Pasha8914** [reply]
This is very strange) because I did an investigation which showed that the official examples behave in the same way as my code

What I found out - multichain works in rainbow, alphawallet, safepal
this cannot work in a trust wallet because they expect a chain id in the transaction parameters


are you sure you used my code without any changes?

**Reply by Pasha8914** [reply]
https://react-app.walletconnect.com/


select polygon and ethereum mainnet here and try to use trustWallet, it won't work either

**Reply by glitch-txs** [reply]
works fine, notice one will ask you for ETH as gas and the other one for MATIC. 

All wallets will expect the chain id as param, but this is handled by ethereum-provider when setting a default chain, which can be handled by the `wallet_switchEthereumChain` method as you did

**Reply by Pasha8914** [reply]
did you use IOS device?

**Reply by glitch-txs** [reply]
Android, maybe try updating the app?

**Comment by finessevanes** [comment]
@Pasha8914 I think this may be related to this [issue](https://github.com/WalletConnect/walletconnect-monorepo/issues/2326). Can you confirm?

**Reply by finessevanes** [reply]
Cool - I'm going to close this discussion then. Please follow this [issue](https://github.com/WalletConnect/walletconnect-monorepo/issues/2326) for updates. Thanks for reporting!

**Reply by thxvatsal** [reply]
Are there any updates on this?

**Reply by liubin595338764** [reply]
hi, Are there any updates on this?


The 2.8.6 version also has the same issue :
```json
{
    "@walletconnect/ethereum-provider": "^2.8.6",
}
```
